### PR TITLE
Extract autoscaling GKE client

### PR DESCRIPTION
**Pull request by aleksandra-malinowska**
Extract autoscaling GKE client from GceManager.

### CODE REVIEW

File: cluster-autoscaler/cloudprovider/gce/autoscaling_gke_client.go [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+	"fmt"
+	"net/http"
+	"time"
+
+	gke "google.golang.org/api/container/v1"
+	gke_beta "google.golang.org/api/container/v1beta1"
+	"k8s.io/autoscaler/cluster-autoscaler/cloudprovider"
+)
+
+// TODO(krzysztof-jastrzebski): Move to main.go.
```
Comment (MaciekPytel) [inline cluster-autoscaler/cloudprovider/gce/autoscaling_gke_client.go]: I would consider removing this TODO. Specifically for this flag it's not bad that it's in a less visible place (it's useless for everyone except GKE developers).

File: cluster-autoscaler/cloudprovider/gce/autoscaling_gke_client.go [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+type NodePool struct {
+	Name              string
+	InstanceGroupUrls []string
+	Autoscaled        bool
+	MinNodeCount      int64
+	MaxNodeCount      int64
+	Autoprovisioned   bool
+}
+
+type autoscalingGkeClientV1 struct {
```
Comment (MaciekPytel) [inline cluster-autoscaler/cloudprovider/gce/autoscaling_gke_client.go]: I don't feel super-strongly about it, but I would put this in autoscaling_gke_client_v1.go (similarly for v1beta1 client).

File: cluster-autoscaler/cloudprovider/gce/autoscaling_gke_client_v1.go [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+	return nodePools, nil
+}
+
+func (m *autoscalingGkeClientV1) FetchLocations() ([]string, error) {
+	cluster, err := m.gkeService.Projects.Locations.Clusters.Get(m.clusterPath).Do()
+	return cluster.Locations, err
+}
+
+func (m *autoscalingGkeClientV1) FetchResourceLimits() (*cloudprovider.ResourceLimiter, error) {
+	return nil, fmt.Errorf("Not implemented")
```
Comment (MaciekPytel) [inline cluster-autoscaler/cloudprovider/gce/autoscaling_gke_client_v1.go]: nit-picking: cloudprovider.ErrNotImplemeted?

File: cluster-autoscaler/cloudprovider/gce/autoscaling_gke_client_v1beta1.go [hunk 4]
```diff
@@ -0,0 +1,193 @@
...
+		if _, found := supportedResources[limit.ResourceType]; !found {
+			glog.Warningf("Unsupported limit defined %s: %d - %d", limit.ResourceType, limit.Minimum, limit.Maximum)
+		}
+		minLimits[limit.ResourceType] = limit.Minimum
+		maxLimits[limit.ResourceType] = limit.Maximum
+	}
+
+	// GKE API provides memory in GB, but ResourceLimiter expects them in bytes
+	if _, found := minLimits[cloudprovider.ResourceNameMemory]; found {
+		minLimits[cloudprovider.ResourceNameMemory] = minLimits[cloudprovider.ResourceNameMemory] * units.Gigabyte
```
Comment (MaciekPytel) [inline cluster-autoscaler/cloudprovider/gce/autoscaling_gke_client_v1beta1.go]: I know it's just copy-paste, but given our recent problems with MB/MiB I wonder if this is the "correct" gigabyte to use?

File: cluster-autoscaler/cloudprovider/gce/autoscaling_gke_client_v1beta1_test.go [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+  "zone": "us-central1-a",
+  "operationType": "CREATE_NODE_POOL",
+  "status": "DONE",
+  "selfLink": "https://container.googleapis.com/v1/projects/601024681890/locations/us-central1-a/operations/operation-1505728466148-d16f5197",
+  "targetLink": "https://container.googleapis.com/v1/projects/601024681890/locations/us-central1-a/clusters/cluster-1/nodePools/nodeautoprovisioning-323233232",
+  "startTime": "2017-09-18T09:54:26.148507311Z",
+  "endTime": "2017-09-18T09:54:35.124878859Z"
+}`
+
+func newTestAutoscalingGkeClientV1beta1(t *testing.T, project, location, clusterName, url string) *autoscalingGkeClientV1beta1 {
```
Comment (MaciekPytel) [inline cluster-autoscaler/cloudprovider/gce/autoscaling_gke_client_v1beta1_test.go]: Shouldn't you override operation timeout & poll interval somewhere in this file?

File: cluster-autoscaler/cloudprovider/gce/gce_manager.go [hunk 6]
```diff
@@ -255,34 +236,18 @@ func CreateGceManager(configReader io.Reader, mode GcpCloudProviderMode, cluster
...
-		gkeBetaService, err := gke_beta.New(client)
+		gkeBetaService, err := NewAutoscalingGkeClientV1beta1(client, projectId, location, clusterName)
 		if err != nil {
 			return nil, err
 		}
-		if *gkeAPIEndpoint != "" {
-			gkeBetaService.BasePath = *gkeAPIEndpoint
-		}
-		manager.gkeBetaService = gkeBetaService
-		glog.V(1).Info("Using GKE-NAP mode")
```
Comment (MaciekPytel) [inline cluster-autoscaler/cloudprovider/gce/gce_manager.go]: I like that comment. Can we keep it?

File: cluster-autoscaler/cloudprovider/gce/gce_manager.go [hunk 7]
```diff
@@ -313,170 +278,29 @@ func (m *gceManagerImpl) assertGCE() {
...
 	changed := false
 
-	for _, nodePool := range nodePoolsResponse.NodePools {
-		autoprovisioned := nodePool.Autoscaling != nil && nodePool.Autoscaling.Autoprovisioned
-		autoscaled := nodePool.Autoscaling != nil && nodePool.Autoscaling.Enabled
-		if !autoscaled {
-			if autoprovisioned {
-				glog.Warningf("NodePool %v has invalid config - autoprovisioned, but not autoscaled. Ignoring this NodePool.", nodePool.Name)
-			}
-			continue
```
Comment (MaciekPytel) [inline cluster-autoscaler/cloudprovider/gce/gce_manager.go]: What's the reason for removing this logic? I'm not sure it was correct to start with, but I'm trying to understand if it's a conscious decision? You could still do the same check using NodePool struct.

Comment (MaciekPytel) [reply cluster-autoscaler/cloudprovider/gce/gce_manager.go]: Looking at it more closely - you don't need this if you filter in FetchNodePools(). But in FetchNodePools() you only check if nodePool.Autoscaling != nil, not if nodePool.Autoscaling.Enabled is true.

File: cluster-autoscaler/cloudprovider/gce/gce_manager.go [hunk 8]
```diff
@@ -576,100 +400,26 @@ func (m *gceManagerImpl) deleteNodePool(toBeRemoved *Mig) error {
...
-	if err != nil {
-		return err
-	}
-	err = m.waitForGkeOp(createOp)
+	err := m.GkeService.CreateNodePool(mig)
 	if err != nil {
 		return err
 	}
-	err = m.fetchAllNodePools()
+	fmt.Println("Create node pool succeeded")
```
Comment (MaciekPytel) [inline cluster-autoscaler/cloudprovider/gce/gce_manager.go]: No println :)

File: cluster-autoscaler/cloudprovider/gce/gce_manager.go [hunk 9]
```diff
@@ -576,100 +400,26 @@ func (m *gceManagerImpl) deleteNodePool(toBeRemoved *Mig) error {
...
+	err = m.refreshNodePools()
 	if err != nil {
 		return err
 	}
+	fmt.Println("Refresh node pool succeeded")
```
Comment (MaciekPytel) [inline cluster-autoscaler/cloudprovider/gce/gce_manager.go]: Delete this one too.

File: cluster-autoscaler/cloudprovider/gce/autoscaling_gke_client.go (hunk 2 above)
Comment (aleksandra-malinowska) [reply cluster-autoscaler/cloudprovider/gce/autoscaling_gke_client.go]: Done

File: cluster-autoscaler/cloudprovider/gce/autoscaling_gke_client_v1beta1_test.go (hunk 5 above)
Comment (aleksandra-malinowska) [reply cluster-autoscaler/cloudprovider/gce/autoscaling_gke_client_v1beta1_test.go]: Done

File: cluster-autoscaler/cloudprovider/gce/gce_manager.go (hunk 6 above)
Comment (aleksandra-malinowska) [reply cluster-autoscaler/cloudprovider/gce/gce_manager.go]: Done

File: cluster-autoscaler/cloudprovider/gce/gce_manager.go (hunk 7 above)
Comment (aleksandra-malinowska) [reply cluster-autoscaler/cloudprovider/gce/gce_manager.go]: Done - added checking Autoscaling.Enabled in FetchNodePools

File: cluster-autoscaler/cloudprovider/gce/gce_manager.go (hunk 8 above)
Comment (aleksandra-malinowska) [reply cluster-autoscaler/cloudprovider/gce/gce_manager.go]: leftover debug printlns, sorry :) should be gone now

File: cluster-autoscaler/cloudprovider/gce/gce_manager.go (hunk 9 above)
Comment (aleksandra-malinowska) [reply cluster-autoscaler/cloudprovider/gce/gce_manager.go]: Done

File: cluster-autoscaler/cloudprovider/gce/autoscaling_gke_client_v1.go (hunk 3 above)
Comment (aleksandra-malinowska) [reply cluster-autoscaler/cloudprovider/gce/autoscaling_gke_client_v1.go]: Done

File: cluster-autoscaler/cloudprovider/gce/autoscaling_gke_client.go (hunk 1 above)
Comment (aleksandra-malinowska) [reply cluster-autoscaler/cloudprovider/gce/autoscaling_gke_client.go]: Done

File: cluster-autoscaler/cloudprovider/gce/gce_manager_test.go [hunk 10]
```diff
@@ -182,39 +180,12 @@ const allNodePools2 = `{
...
-      },
-      "initialNodeCount": 3,
-      "autoscaling": {},
-      "management": {},
-      "selfLink": "https://container.googleapis.com/v1/projects/project1/zones/us-central1-b/clusters/cluster-1/nodePools/node_pool3",
-      "version": "1.6.9",
-      "instanceGroupUrls": [
-        "https://www.googleapis.com/compute/v1/projects/project1/zones/us-central1-b/instanceGroupManagers/gke-cluster-1-node_pool3"
-      ],
-      "status": "RUNNING"
```
[the remark this answers is not in the document]
Comment (aleksandra-malinowska) [reply cluster-autoscaler/cloudprovider/gce/gce_manager_test.go]: If you look closely at this json, default-pool is listed twice. That looks neither correct nor intentional.

File: cluster-autoscaler/cloudprovider/gce/autoscaling_gke_client_v1beta1.go (hunk 4 above)
Comment (aleksandra-malinowska) [reply cluster-autoscaler/cloudprovider/gce/autoscaling_gke_client_v1beta1.go]: It's 2^30. Whether this is what we want to do or not is another question :) Either way, it seems to be an API decision and outside the scope of refactoring.

File: cluster-autoscaler/cloudprovider/gce/gce_manager_test.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (MaciekPytel) [reply cluster-autoscaler/cloudprovider/gce/gce_manager_test.go]: Looking closely at those tests is something I'm trying to avoid as much as possible :s

**Comment by MaciekPytel** [comment]
Please squash the commits before merging (and I'll re-lgtm)

**Pull request merged**